### PR TITLE
Use describe for generated controller tests

### PR DESCRIPTION
**Pull request by Gazler**
Opening for discussion as suggested by @chrismccord 

Here is an example gist containing the new and old generated controller tests:

new: https://gist.github.com/Gazler/aa4b6dfcd92cbe86e58e4ad22918b4fe#file-new_controller_test-exs
old: https://gist.github.com/Gazler/aa4b6dfcd92cbe86e58e4ad22918b4fe#file-old_controller_test-exs

If we want to continue with this, I'll update the rest of the generators.

### CODE REVIEW

File: priv/templates/phx.gen.html/controller_test.exs [hunk 1]
```diff
@@ -12,59 +12,78 @@ defmodule <%= inspect context.web_module %>.<%= inspect Module.concat(schema.web
...
+    test "lists all entries on index", %{conn: conn} do
+      conn = get conn, <%= schema.route_helper %>_path(conn, :index)
+      assert html_response(conn, 200) =~ "Listing <%= schema.human_plural %>"
+    end
   end
 
-  test "renders form for new <%= schema.plural %>", %{conn: conn} do
-    conn = get conn, <%= schema.route_helper %>_path(conn, :new)
-    assert html_response(conn, 200) =~ "New <%= schema.human_singular %>"
+  describe "new" do
```
Comment (josevalim) [inline priv/templates/phx.gen.html/controller_test.exs]: What if we change the wording on those?

```elixir
describe "new <%= schema.singular %>" do
  test "renders input form"
end

describe "create <%= schema.plural %>" do
  test "redirects to show when data is valid"
  test "render errors when data is invalid"
end
```

etc?

Comment (Gazler) [inline priv/templates/phx.gen.html/controller_test.exs]: The intention here was to keep them matching the action names. Am happy to append the resource name though. I'll update the PR.

**Comment by Gazler** [comment]
HTML
```elixir
defmodule Foo.Web.UserControllerTest do
  use Foo.Web.ConnCase

  alias Foo.Accounts

  @create_attrs %{age: 42, email: "some email", name: "some name"}
  @update_attrs %{age: 43, email: "some updated email", name: "some updated name"}
  @invalid_attrs %{age: nil, email: nil, name: nil}

  def fixture(:user) do
    {:ok, user} = Accounts.create_user(@create_attrs)
    user
  end

  describe "index" do
    test "lists all entries on index", %{conn: conn} do
      conn = get conn, user_path(conn, :index)
      assert html_response(conn, 200) =~ "Listing Users"
    end
  end

  describe "new user" do
    test "renders form for new users", %{conn: conn} do
      conn = get conn, user_path(conn, :new)
      assert html_response(conn, 200) =~ "New User"
    end
  end

  describe "create user" do
    test "creates user and redirects to show when data is valid", %{conn: conn} do
      conn = post conn, user_path(conn, :create), user: @create_attrs

      assert %{id: id} = redirected_params(conn)
      assert redirected_to(conn) == user_path(conn, :show, id)

      conn = get conn, user_path(conn, :show, id)
      assert html_response(conn, 200) =~ "Show User"
    end

    test "does not create user and renders errors when data is invalid", %{conn: conn} do
      conn = post conn, user_path(conn, :create), user: @invalid_attrs
      assert html_response(conn, 200) =~ "New User"
    end
  end

  describe "edit user" do
    setup [:create_user]

    test "renders form for editing chosen user", %{conn: conn, user: user} do
      conn = get conn, user_path(conn, :edit, user)
      assert html_response(conn, 200) =~ "Edit User"
    end
  end

  describe "update user" do
    setup [:create_user]

    test "updates chosen user and redirects when data is valid", %{conn: conn, user: user} do
      conn = put conn, user_path(conn, :update, user), user: @update_attrs
      assert redirected_to(conn) == user_path(conn, :show, user)

      conn = get conn, user_path(conn, :show, user)
      assert html_response(conn, 200) =~ "some updated email"
    end

    test "does not update chosen user and renders errors when data is invalid", %{conn: conn, user: user} do
      conn = put conn, user_path(conn, :update, user), user: @invalid_attrs
      assert html_response(conn, 200) =~ "Edit User"
    end
  end

  describe "delete user" do
    setup [:create_user]

    test "deletes chosen user", %{conn: conn, user: user} do
      conn = delete conn, user_path(conn, :delete, user)
      assert redirected_to(conn) == user_path(conn, :index)
      assert_error_sent 404, fn ->
        get conn, user_path(conn, :show, user)
      end
    end
  end

  defp create_user(_) do
    user = fixture(:user)
    {:ok, user: user}
  end
end
```

JSON:
```elixir
defmodule Foo.Web.UserControllerTest do
  use Foo.Web.ConnCase

  alias Foo.Accounts
  alias Foo.Accounts.User

  @create_attrs %{age: 42, email: "some email", name: "some name"}
  @update_attrs %{age: 43, email: "some updated email", name: "some updated name"}
  @invalid_attrs %{age: nil, email: nil, name: nil}

  def fixture(:user) do
    {:ok, user} = Accounts.create_user(@create_attrs)
    user
  end

  setup %{conn: conn} do
    {:ok, conn: put_req_header(conn, "accept", "application/json")}
  end

  describe "index" do
    test "lists all entries on index", %{conn: conn} do
      conn = get conn, user_path(conn, :index)
      assert json_response(conn, 200)["data"] == []
    end
  end

  describe "create user" do
    test "creates user and renders user when data is valid", %{conn: conn} do
      conn = post conn, user_path(conn, :create), user: @create_attrs
      assert %{"id" => id} = json_response(conn, 201)["data"]

      conn = get conn, user_path(conn, :show, id)
      assert json_response(conn, 200)["data"] == %{
        "id" => id,
        "age" => 42,
        "email" => "some email",
        "name" => "some name"}
    end

    test "does not create user and renders errors when data is invalid", %{conn: conn} do
      conn = post conn, user_path(conn, :create), user: @invalid_attrs
      assert json_response(conn, 422)["errors"] != %{}
    end
  end

  describe "update user" do
    setup [:create_user]

    test "updates chosen user and renders user when data is valid", %{conn: conn, user: %User{id: id} = user} do
      conn = put conn, user_path(conn, :update, user), user: @update_attrs
      assert %{"id" => ^id} = json_response(conn, 200)["data"]

      conn = get conn, user_path(conn, :show, id)
      assert json_response(conn, 200)["data"] == %{
        "id" => id,
        "age" => 43,
        "email" => "some updated email",
        "name" => "some updated name"}
    end

    test "does not update chosen user and renders errors when data is invalid", %{conn: conn, user: user} do
      conn = put conn, user_path(conn, :update, user), user: @invalid_attrs
      assert json_response(conn, 422)["errors"] != %{}
    end
  end

  describe "delete user" do
    setup [:create_user]

    test "deletes chosen user", %{conn: conn, user: user} do
      conn = delete conn, user_path(conn, :delete, user)
      assert response(conn, 204)
      assert_error_sent 404, fn ->
        get conn, user_path(conn, :show, user)
      end
    end
  end

  defp create_user(_) do
    user = fixture(:user)
    {:ok, user: user}
  end
end
```

**Comment by josevalim** [comment]
@Gazler i am not a fan of the "create user creates user" - I would remove the second one as I originally proposed. But I guess it depends if the describe should be about the behaviour or the action. Also don't worry about squashing for now in case we decide to revert the changes.

**Comment by Gazler** [comment]
@josevalim What about:

```elixir
describe "index"
  test "lists all entries"

describe "new user"
  test "renders form"

describe "create user"
  test "redirects to show when data is valid"
  test "renders errors when data is invalid"

describe "edit user"
  test "renders form for editing chosen user"

describe "update user"
  test "updates and redirects when data is valid"
  test "does not update the user and renders errors when data is invalid"

describe "delete user"
  test "deletes the chosen user"
```

**Comment by chrismccord** [comment]
I like these changes after seeing them side by side, especially the example of the `:create_user` shared setup. I'm good with the above e wording unless @josevalim has better ideas?

**Comment by josevalim** [comment]
@chrismccord @Gazler there is an inconsistency in create/update. create does not repeat the action, update does. We need to pick one the other. I won't get involved since i am not a native speaker but we should be as concise and clear as possible.

**Comment by Gazler** [comment]
I think:

```
describe "update user"
  test "redirects when data is valid"
```

Is probably sufficient to convey that the user is actually updated instead of just doing a redirect.

**Comment by chrismccord** [comment]
👍

**Comment by josevalim** [comment]
:shipit:

**Pull request merged**